### PR TITLE
[AUTO] Exclude other vendor's GPU device in default candidate list

### PR DESCRIPTION
**Pull request by WeldonWangwang**
### Details:
 - [AUTO] Exclude other vendor's GPU device in default candidate list
 - User can try other vendor's GPU device by setting ov::device::priorities explicitly.

### Tickets:
 - *CVS-106628*

### CODE REVIEW

File: src/plugins/auto/src/plugin.cpp [hunk 1]
```diff
@@ -549,17 +560,22 @@ std::list<DeviceInformation> MultiDeviceInferencePlugin::GetValidDevice(
...
+                }
+            } else {
+                otherGPU.push_back(item);
             }
             continue;
         }
     }
 
-    // Priority of selecting device: dGPU > VPUX > iGPU > MYRIAD > CPU
+    // Priority of selecting device: dGPU > VPUX > iGPU > MYRIAD > CPU > otherGPU
```
Comment (WeldonWangwang) [inline src/plugins/auto/src/plugin.cpp:578]: @peterchen-intel Is this the device priority we are expected？

File: src/plugins/auto/src/plugin.cpp [hunk 2]
```diff
@@ -528,6 +528,18 @@ std::list<DeviceInformation> MultiDeviceInferencePlugin::GetValidDevice(
     if (metaDevices.empty()) {
         IE_THROW(NotFound) << "No available device to select in " << GetName() << " plugin";
     }
+    auto isIntelDevice = [](std::string architectureInfo) {
+        std::string startString = "GPU: vendor=";
+        std::string endString = " arch=";
+        auto start = architectureInfo.find(startString);
+        auto end = architectureInfo.find(endString);
+        auto vendorID = architectureInfo.substr(start + startString.length(), end - start - startString.length());
+        if (vendorID == "0x8086") {
```
Comment (yangwang201911) [inline src/plugins/auto/src/plugin.cpp:537]:
```suggestion
        return vendorID == "0x8086";
```

File: src/plugins/auto/src/plugin.cpp [hunk 3]
```diff
@@ -549,11 +561,19 @@ std::list<DeviceInformation> MultiDeviceInferencePlugin::GetValidDevice(
             continue;
         }
         if (item.deviceName.find("GPU") == 0) {
-            auto& gpuUniqueName = item.uniqueName;
-            if (gpuUniqueName.find("iGPU") != std::string::npos) {
-                iGPU.push_back(item);
-            } else if (gpuUniqueName.find("dGPU") != std::string::npos) {
-                dGPU.push_back(item);
+            auto deviceArchitecture = GetCore()->GetMetric(item.deviceName, METRIC_KEY(DEVICE_ARCHITECTURE)).as<std::string>();
+            if (isIntelDevice(deviceArchitecture)) {
```
Comment (yangwang201911) [inline src/plugins/auto/src/plugin.cpp:565]: Does this logic mean we just only accept GPU from INTEL? Is this our expected behavior?

Comment (WeldonWangwang) [reply src/plugins/auto/src/plugin.cpp:565]: From the discussion in cvs-106628, will not add other GPU to device priorities. Please check this @peterchen-intel @wangleis

File: src/plugins/auto/src/plugin.cpp (hunk 2 above)
Comment (WeldonWangwang) [reply src/plugins/auto/src/plugin.cpp:537]: Okay.

File: src/plugins/auto/src/plugin.cpp (hunk 3 above)
Comment (wangleis) [reply src/plugins/auto/src/plugin.cpp:565]: Let's select Intel dGPU as first priority and keep other dGPU as low priority which just higher than CPU.

File: src/plugins/auto/src/plugin.cpp [hunk 4]
```diff
@@ -528,6 +528,14 @@ std::list<DeviceInformation> MultiDeviceInferencePlugin::GetValidDevice(
     if (metaDevices.empty()) {
         IE_THROW(NotFound) << "No available device to select in " << GetName() << " plugin";
     }
+    auto isIntelDevice = [](std::string architectureInfo) {
+        std::string startString = "GPU: vendor=";
```
[the remark this answers is not in the document]
Comment (WeldonWangwang) [reply src/plugins/auto/src/plugin.cpp:532]: Agree.

**Comment by ilya-lavrenov** [comment]
is it a target for 2023.0 ?

**Comment by peterchen-intel** [comment]
> is it a target for 2023.0 ?

I assume it is a nice to have. @nkogteva can you comment?

**Comment by nkogteva** [comment]
> > is it a target for 2023.0 ?
> 
> I assume it is a nice to have. @nkogteva can you comment?

@ilya-lavrenov @peterchen-intel  I think so, otherwise user who have nvidia can face situation when AUTO plugin selects NVIDIA GPU under Intel GPU plugin.

**Review by wangleis — APPROVED** [review]

**Comment by wangleis** [comment]
@ilya-lavrenov @nkogteva Just for clarification on this PR. AUTO will not use NVIDIA when user just use `AUTO`. But AUTO will use all GPU devices when user use `AUTO:GPU` or `AUTO:GPU,CPU`.

File: src/plugins/auto/src/plugin.cpp [hunk 5]
```diff
@@ -674,14 +679,21 @@ void MultiDeviceInferencePlugin::RegisterPriority(const unsigned int& priority,
...
         // filter out the supported devices
-        if (!_pluginConfig.isSupportedDevice(device))
+        if (device.find("GPU") != std::string::npos) {
+            deviceArchitecture = GetCore()->GetMetric(device, METRIC_KEY(DEVICE_ARCHITECTURE)).as<std::string>();
+        }
+        if (!_pluginConfig.isSupportedDevice(device, deviceArchitecture))
             continue;
         allDevices += device + ",";
     }
+    if (allDevices.empty()) {
```
Comment (yangwang201911) [inline src/plugins/auto/src/plugin.cpp:694]: we also need to revert this logic into the end of this function as before as AUTO should throw this exception for the case `-d AUTO:-CPU,-GPU.0,-GPU.1`. Maybe we need additional test case here to check if the exception throwed as we expected in this function.

File: src/plugins/auto/tests/unit/get_device_list.cpp [hunk 6]
```diff
@@ -113,6 +140,11 @@ const std::vector<Params> testConfigsWithId = {Params{" ", " "},
                                          Params{"-GPU.0,-GPU.1", "CPU"},
                                          Params{"-GPU.0,-GPU.1,INVALID_DEVICE", "INVALID_DEVICE"},
                                          Params{"-GPU.0,-GPU.1,-INVALID_DEVICE", "CPU"},
+                                         Params{"-GPU.0,-GPU.1,-CPU", ""},
```
Comment (yangwang201911) [inline src/plugins/auto/tests/unit/get_device_list.cpp:143]: maybe need to add this case into the test cases for throwing exception test.

File: src/plugins/auto/tests/unit/get_device_list.cpp [hunk 7]
```diff
@@ -139,6 +171,36 @@ const std::vector<Params> testConfigs = {Params{" ", " "},
...
+                                         Params{"CPU,GPU,VPUX,INVALID_DEVICE", "CPU,GPU.0,GPU.1,VPUX,INVALID_DEVICE"},
+                                         Params{"VPUX,GPU,CPU,-GPU.0", "VPUX,GPU.1,CPU"},
+                                         Params{"-GPU.0,GPU,CPU", "GPU.1,CPU"},
+                                         Params{"-GPU.0,GPU", "GPU.1"},
+                                         Params{"-GPU,GPU.0", "GPU.0"},
+                                         Params{"-GPU.0", "CPU"},
+                                         Params{"-GPU.0,-GPU.1", "CPU"},
+                                         Params{"-GPU.0,-GPU.1,INVALID_DEVICE", "INVALID_DEVICE"},
+                                         Params{"-GPU.0,-GPU.1,-INVALID_DEVICE", "CPU"},
+                                         Params{"-GPU.0,-GPU.1,-CPU", ""},
```
Comment (yangwang201911) [inline src/plugins/auto/tests/unit/get_device_list.cpp:191]: same as above comment.

File: src/plugins/auto/src/plugin.cpp [hunk 8]
```diff
@@ -549,10 +549,15 @@ std::list<DeviceInformation> MultiDeviceInferencePlugin::GetValidDevice(
...
         }
         if (item.deviceName.find("GPU") == 0) {
-            auto& gpuUniqueName = item.uniqueName;
-            if (gpuUniqueName.find("iGPU") != std::string::npos) {
+            ov::Any deviceType;
+            if (IsNewAPI()) {
+                deviceType = GetCore()->GetMetric(item.deviceName, METRIC_KEY(DEVICE_TYPE)).as<ov::device::Type>();
+            } else {
+                deviceType = GetCore()->GetMetric(item.deviceName, METRIC_KEY(DEVICE_TYPE)).as<Metrics::DeviceType>();
+            }
```
Comment (songbell) [inline src/plugins/auto/src/plugin.cpp:557]: why still need to check new API, device type is mapped in properties already:

/** @cond INTERNAL */
inline std::ostream& operator<<(std::ostream& os, const InferenceEngine::Metrics::DeviceType& deviceType) {
    switch (deviceType) {
    case InferenceEngine::Metrics::DeviceType::discrete:
        os << "discrete";
        break;
    case InferenceEngine::Metrics::DeviceType::integrated:
        os << "integrated";
        break;
    default:
        os << "unknown";
        break;
    }

    return os;
}

Comment (WeldonWangwang) [reply src/plugins/auto/src/plugin.cpp:557]: Okay.

**Review by yangwang201911 — APPROVED** [review]

File: src/plugins/auto/src/plugin.cpp [hunk 9]
```diff
@@ -708,13 +723,32 @@ std::string MultiDeviceInferencePlugin::GetDeviceList(const std::map<std::string
             });
             return iter != devices.end();
         };
+        auto isAnyDevWithEmptyMerged = [](std::string& device, const std::vector<std::string>& devices) {
+            auto iter = std::find_if(devices.begin(), devices.end(), [device](const std::string& devItem) {
+                std::string _device = device;
```
Comment (songbell) [inline src/plugins/auto/src/plugin.cpp:728]: minor: not use _ for non member variables.

File: src/plugins/auto/src/plugin.cpp [hunk 10]
```diff
@@ -708,13 +723,32 @@ std::string MultiDeviceInferencePlugin::GetDeviceList(const std::map<std::string
...
+            return iter != devices.end();
+        };
         auto deviceWithDefaultID = [](std::string& device) {
             // AUTO assume the default device ID will be "0" for the single device.
             return device.find(".") == std::string::npos ? device + ".0" : device;
         };
         if (devicesToBeMerged.empty()) {
             for (auto&& device : deviceList) {
-                if (isAnyDev(device, devicesToBeDeleted) || !_pluginConfig.isSupportedDevice(device))
+                if (device.find("GPU") != std::string::npos) {
```
Comment (songbell) [inline src/plugins/auto/src/plugin.cpp:743]: let's optimize after, these lines are duplicated with above lines #689

File: src/plugins/auto/src/plugin_config.hpp [hunk 11]
```diff
@@ -183,14 +183,20 @@ class PluginConfig {
         return pluginName == "AUTO" ? supported_metrics : multi_supported_metrics;
     }
 
-    bool isSupportedDevice(const std::string& deviceName) const {
+    bool isSupportedDevice(const std::string& deviceName, std::string option) const {
```
Comment (songbell) [inline src/plugins/auto/src/plugin_config.hpp:186]: minor : consider const &

**Review by songbell — APPROVED** [review]

File: src/plugins/auto/src/plugin.cpp (hunk 10 above)
Comment (WeldonWangwang) [reply src/plugins/auto/src/plugin.cpp:748]: Optimized.

File: src/plugins/auto/src/plugin_config.hpp (hunk 11 above)
Comment (WeldonWangwang) [reply src/plugins/auto/src/plugin_config.hpp]: Done.

File: src/plugins/auto/src/plugin.cpp [hunk 12]
```diff
@@ -674,11 +681,24 @@ void MultiDeviceInferencePlugin::RegisterPriority(const unsigned int& priority,
...
     std::string allDevices;
+    std::string deviceArchitecture;
     auto deviceList = GetCore()->GetAvailableDevices();
     auto deviceListConfig = config.find(ov::device::priorities.name());
+    auto getGpuArchitecture = [&](const std::string& name) -> std::string {
+        try {
+            auto architectureInfo = GetCore()->GetMetric(name, METRIC_KEY(DEVICE_ARCHITECTURE)).as<std::string>();
+            return architectureInfo;
+        } catch (const IE::Exception&) {
+            LOG_DEBUG_TAG("GetMetric:%s for %s failed ", "DEVICE_ARCHITECTURE", name.c_str());
```
[the remark this answers is not in the document]
Comment (peterchen-intel) [reply src/plugins/auto/src/plugin.cpp:692]: will go to next statement and return "".

**Pull request merged**